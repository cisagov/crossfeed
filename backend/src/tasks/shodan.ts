import { Domain, Service, Vulnerability } from '../models';
import { plainToClass } from 'class-transformer';
import getIps from './helpers/getIps';
import { CommandOptions } from './ecs-client';
import saveServicesToDb from './helpers/saveServicesToDb';
import { chunk } from 'lodash';
import axios from 'axios';
import pRetry from 'p-retry';
import { IpToDomainsMap, sanitizeStringField } from './censysIpv4';
import saveVulnerabilitiesToDb from './helpers/saveVulnerabilitiesToDb';

// Shodan allows searching up to 100 IPs at once
const CHUNK_SIZE = 100;

interface ShodanResponse {
  ip_str: string;
  country_name: string;
  domains: string[];
  hostnames: string[];
  org: string;
  ports: number[];
  os: string;
  city: string;
  longitude: number;
  latitude: number;
  data: {
    port: number;
    product: string;
    data: string;
    cpe: string[];
    version: string;
    vulns: {
      [title: string]: {
        verified: boolean;
        references: string[];
        cvss: number;
        summary: string;
      };
    };
  }[];
}

const sleep = (milliseconds) => {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
};

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running shodan on organization', organizationName);

  const domainsWithIPs = await getIps(organizationId);

  const ipToDomainsMap: IpToDomainsMap = domainsWithIPs.reduce<IpToDomainsMap>(
    (map: IpToDomainsMap, domain: Domain) => {
      if (!map[domain.ip]) {
        map[domain.ip] = [];
      }
      map[domain.ip].push(domain);
      return map;
    },
    {}
  );

  const chunks = chunk(domainsWithIPs, CHUNK_SIZE);

  for (const domainChunk of chunks) {
    console.log(
      `Scanning ${domainChunk.length} domains beginning with ${domainChunk[0].name}`
    );
    try {
      let { data } = await pRetry(
        () =>
          axios.get<ShodanResponse[]>(
            `https://api.shodan.io/shodan/host/${encodeURI(
              domainChunk.map((domain) => domain.ip).join(',')
            )}?key=${process.env.SHODAN_API_KEY}`
          ),
        {
          // Perform less retries on jest to make tests faster
          retries: typeof jest === 'undefined' ? 5 : 2,
          minTimeout: 1000
        }
      );

      // If only one item is returned, the response will not be an array
      if (!Array.isArray(data)) {
        data = [data];
      }
      for (const item of data) {
        const domains = ipToDomainsMap[item.ip_str];
        for (const domain of domains) {
          for (const service of item.data) {
            const [serviceId] = await saveServicesToDb([
              plainToClass(Service, {
                domain: domain,
                discoveredBy: { id: commandOptions.scanId },
                port: service.port,
                lastSeen: new Date(Date.now()),
                banner: sanitizeStringField(service.data),
                shodanResults: {
                  product: service.product,
                  version: service.version,
                  cpe: service.cpe
                }
              })
            ]);
            if (service.vulns) {
              const vulns: Vulnerability[] = [];
              for (const cve in service.vulns) {
                // console.log('Creating vulnerability', cve);
                vulns.push(
                  plainToClass(Vulnerability, {
                    domain: domain,
                    lastSeen: new Date(Date.now()),
                    title: cve,
                    cve: cve,
                    // Shodan CPE information is unreliable,
                    // so don't add it in for now.
                    // cpe:
                    //   service.cpe && service.cpe.length > 0
                    //     ? service.cpe[0]
                    //     : null,
                    cvss: service.vulns[cve].cvss,
                    state: 'open',
                    source: 'shodan',
                    needsPopulation: true,
                    service: { id: serviceId }
                  })
                );
              }
              await saveVulnerabilitiesToDb(vulns, false);
            }
          }
        }
      }
    } catch (e) {
      console.error(e);
    }

    await sleep(1000); // Wait for Shodan rate limit of 1 request / second
  }

  console.log(`Shodan finished for ${domainsWithIPs.length} domains`);
};
