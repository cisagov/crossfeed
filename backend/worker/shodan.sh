#!/bin/bash

set -e

cd /app/pe-reports

echo "Starting Shodan"

pe-source shodan --orgs=DHS --soc_med_included

echo "Done"