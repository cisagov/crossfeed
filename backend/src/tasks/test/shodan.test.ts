import * as nock from 'nock';
import {
  connectToDatabase,
  Domain,
  Organization,
  Scan,
  Service,
  Vulnerability
} from '../../models';
import { handler as shodan } from '../shodan';

const RealDate = Date;

const shodanResponse = [
  {
    region_code: null,
    ip: 2575209532,
    postal_code: null,
    country_code: 'JP',
    city: null,
    dma_code: null,
    last_update: '2020-12-16T20:00:25.339321',
    latitude: 35.69,
    tags: ['self-signed', 'starttls'],
    area_code: null,
    country_name: 'Japan',
    hostnames: ['otakukonkatsu.com'],
    org: 'SAKURA Internet',
    data: [
      {
        hash: -1790423402,
        os: null,
        tags: ['self-signed'],
        opts: {
          vulns: [],
          heartbleed: '2020/12/16 20:00:12 153.126.148.60:993 - SAFE\n'
        },
        ip: 2575209532,
        isp: 'SAKURA Internet',
        port: 993,
        hostnames: ['otakukonkatsu.com'],
        timestamp: '2020-12-16T20:00:25.339321',
        domains: ['otakukonkatsu.com'],
        org: 'SAKURA Internet',
        data:
          '* OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE AUTH=PLAIN AUTH=LOGIN] Dovecot ready.\n* CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE AUTH=PLAIN AUTH=LOGIN\r\nA001 OK Capability completed.\r\n* ID NIL\r\nA002 OK ID completed.\r\nA003 BAD Error in IMAP command received by server.\r\n',
        asn: 'AS7684',
        transport: 'tcp',
        ip_str: '153.126.148.60',
        product: 'Test',
        version: '1.1',
        cpe: ['cpe:/a:igor_sysoev:nginx:1.18.0', 'cpe:/a:atlassian:confluence']
      }
    ],
    asn: 'AS7684',
    isp: 'SAKURA Internet',
    longitude: 139.69,
    country_code3: null,
    domains: ['otakukonkatsu.com'],
    ip_str: '153.126.148.60',
    os: null,
    ports: [993, 587, 110, 80, 465, 25, 443]
  },
  {
    region_code: null,
    ip: 16843009,
    postal_code: null,
    country_code: 'AU',
    city: null,
    dma_code: null,
    last_update: '2020-12-21T22:04:16.508070',
    latitude: -33.494,
    tags: [],
    area_code: null,
    country_name: 'Australia',
    hostnames: ['one.one.one.one'],
    org: 'Mountain View Communications',
    data: [
      {
        _shodan: {
          id: '8eb5fb82-81c8-4a6d-bc20-43d83079069b',
          options: {},
          ptr: true,
          module: 'dns-udp',
          crawler: '70752434fdf0dcec35df6ae02b9703eaae035f7d'
        },
        hash: 1592421393,
        os: null,
        opts: {
          raw: '34ef818500010000000000000776657273696f6e0462696e640000100003'
        },
        ip: 16843009,
        isp: 'Mountain View Communications',
        port: 53,
        hostnames: ['one.one.one.one'],
        location: {
          city: null,
          region_code: null,
          area_code: null,
          longitude: 143.2104,
          country_code3: null,
          country_name: 'Australia',
          postal_code: null,
          dma_code: null,
          country_code: 'AU',
          latitude: -33.494
        },
        dns: {
          resolver_hostname: null,
          recursive: true,
          resolver_id: 'AMS',
          software: null
        },
        timestamp: '2020-12-21T22:04:16.508070',
        domains: ['one.one'],
        org: 'Mountain View Communications',
        data: '\nRecursion: enabled\nResolver ID: AMS',
        asn: 'AS13335',
        transport: 'udp',
        ip_str: '1.1.1.1',
        vulns: {
          'CVE-1234-1234': {
            verified: true,
            references: [],
            cvss: '5',
            summary: 'A vulnerability'
          }
        }
      }
    ],
    asn: 'AS13335',
    isp: 'Mountain View Communications',
    longitude: 143.2104,
    country_code3: null,
    domains: ['one.one'],
    ip_str: '1.1.1.1',
    os: null,
    ports: [53]
  }
];

describe('shodan', () => {
  let organization;
  let scan;
  const domains: Domain[] = [];
  beforeEach(async () => {
    await connectToDatabase();
    global.Date.now = jest.fn(() => new Date('2019-04-22T10:20:30Z').getTime());
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
    scan = await Scan.create({
      name: 'shodan',
      arguments: {},
      frequency: 999
    }).save();
    domains.push(
      await Domain.create({
        name: 'first_file_testdomain1',
        ip: '153.126.148.60',
        organization
      }).save()
    );
    domains.push(
      await Domain.create({
        name: 'first_file_testdomain2',
        ip: '31.134.10.156',
        organization
      }).save()
    );
    domains.push(
      await Domain.create({
        name: 'first_file_testdomain12',
        ip: '1.1.1.1',
        organization
      }).save()
    );

    jest.mock('../helpers/getIps', () => domains);
  });

  afterEach(async () => {
    global.Date = RealDate;
  });

  const checkDomains = async (organization) => {
    const domains = await Domain.find({
      where: { organization },
      relations: ['organization', 'services']
    });
    expect(
      domains
        .sort((a, b) => a.name.localeCompare(b.name))
        .map((e) => ({
          ...e,
          services: e.services.map((s) => ({
            ...s,
            id: null,
            updatedAt: null,
            createdAt: null
          })),
          organization: null,
          id: null,
          updatedAt: null,
          createdAt: null,
          syncedAt: null,
          name: null
        }))
    ).toMatchSnapshot();
    expect(domains.filter((e) => !e.organization).length).toEqual(0);
  };
  test('basic test', async () => {
    nock('https://api.shodan.io')
      .get(
        `/shodan/host/153.126.148.60,31.134.10.156,1.1.1.1?key=${process.env.SHODAN_API_KEY}`
      )
      .reply(200, shodanResponse);
    await shodan({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    await checkDomains(organization);
  });
  test.skip('creates vulnerability', async () => {
    nock('https://api.shodan.io')
      .get(
        `/shodan/host/153.126.148.60,31.134.10.156,1.1.1.1?key=${process.env.SHODAN_API_KEY}`
      )
      .reply(200, shodanResponse);
    await shodan({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    const domain = await Domain.findOne({ id: domains[2].id });
    const vulns = await Vulnerability.find({
      domain: domain
    });
    expect(vulns).toHaveLength(1);
    expect(vulns[0].title).toEqual('CVE-1234-1234');
  });
  test('populates shodanResults and products', async () => {
    nock('https://api.shodan.io')
      .get(
        `/shodan/host/153.126.148.60,31.134.10.156,1.1.1.1?key=${process.env.SHODAN_API_KEY}`
      )
      .reply(200, shodanResponse);
    await shodan({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    const service = await Service.findOne({ domain: domains[0], port: 993 });
    expect(service).not.toBeUndefined();
    expect(service!.shodanResults).toEqual({
      product: 'Test',
      version: '1.1',
      cpe: ['cpe:/a:igor_sysoev:nginx:1.18.0', 'cpe:/a:atlassian:confluence']
    });
    expect(service!.products).toHaveLength(2);
    expect(service!.products).toEqual([
      {
        cpe: 'cpe:/a:igor_sysoev:nginx:1.18.0',
        name: 'nginx',
        tags: [],
        vendor: 'igor sysoev',
        version: '1.18.0'
      },
      {
        cpe: 'cpe:/a:atlassian:confluence',
        name: 'confluence',
        tags: [],
        vendor: 'atlassian'
      }
    ]);
  });
});
