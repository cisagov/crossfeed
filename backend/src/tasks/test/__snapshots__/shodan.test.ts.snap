// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shodan basic test 1`] = `
Array [
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "ipOnly": false,
    "name": null,
    "organization": null,
    "reverseName": "first_file_testdomain1",
    "screenshot": null,
    "services": Array [
      Object {
        "banner": "* OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE AUTH=PLAIN AUTH=LOGIN] Dovecot ready.
* CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE AUTH=PLAIN AUTH=LOGIN
A001 OK Capability completed.
* ID NIL
A002 OK ID completed.
A003 BAD Error in IMAP command received by server.
",
        "censysIpv4Results": Object {},
        "censysMetadata": Object {},
        "createdAt": null,
        "id": null,
        "intrigueIdentResults": Object {},
        "lastSeen": 2019-04-22T10:20:30.000Z,
        "port": 993,
        "productSource": null,
        "products": Array [
          Object {
            "cpe": "cpe:/a:igor_sysoev:nginx:1.18.0",
            "name": "nginx",
            "tags": Array [],
            "vendor": "igor sysoev",
            "version": "1.18.0",
          },
          Object {
            "cpe": "cpe:/a:atlassian:confluence",
            "name": "confluence",
            "tags": Array [],
            "vendor": "atlassian",
          },
        ],
        "service": null,
        "serviceSource": "shodan",
        "shodanResults": Object {
          "cpe": Array [
            "cpe:/a:igor_sysoev:nginx:1.18.0",
            "cpe:/a:atlassian:confluence",
          ],
          "product": "Test",
          "version": "1.1",
        },
        "updatedAt": null,
        "wappalyzerResults": Array [],
      },
    ],
    "ssl": null,
    "subdomainSource": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "1.1.1.1",
    "ipOnly": false,
    "name": null,
    "organization": null,
    "reverseName": "first_file_testdomain12",
    "screenshot": null,
    "services": Array [
      Object {
        "banner": "
Recursion: enabled
Resolver ID: AMS",
        "censysIpv4Results": Object {},
        "censysMetadata": Object {},
        "createdAt": null,
        "id": null,
        "intrigueIdentResults": Object {},
        "lastSeen": 2019-04-22T10:20:30.000Z,
        "port": 53,
        "productSource": null,
        "products": Array [],
        "service": null,
        "serviceSource": "shodan",
        "shodanResults": Object {},
        "updatedAt": null,
        "wappalyzerResults": Array [],
      },
    ],
    "ssl": null,
    "subdomainSource": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "ipOnly": false,
    "name": null,
    "organization": null,
    "reverseName": "first_file_testdomain2",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "subdomainSource": null,
    "syncedAt": null,
    "updatedAt": null,
  },
]
`;
